README.md
[![Build Status](https://travis-ci.org/pattonwebz/best-reloaded-bootstrap-theme.svg?branch=master)](https://travis-ci.org/pattonwebz/best-reloaded-bootstrap-theme) [![Build Status](https://scrutinizer-ci.com/g/pattonwebz/best-reloaded-bootstrap-theme/badges/build.png?b=master)](https://scrutinizer-ci.com/g/pattonwebz/best-reloaded-bootstrap-theme/build-status/master)

# Using The Best Reloaded Theme Package

This theme uses NPM and Grunt to run various development and build tasks. To get started you will need to install the dependencies. You may already have grunt-cli installed globally.

```
npm install grunt-cli -g
npm install
```

Then you can run the Grunt tasks. The initial build task handles copying of the Bootstrap 4 sourcefiles (they are installed during 'npm install') into the theme from the node_modules directory. It will also process them and finally it will combine & minify the Bootstrap, and the main themes, styles/scripts.

## All Available Grunt Tasks

```
# This is the initial build task. It copies source files from inside 'node_modules/' into the theme and processes all styles and scripts.
# The full build task exists primarily to deal with Bootstrap files. You must run this once to pull Bootstrap source files into the directory and process them.
# After that you will only need to run this again if Bootstrap source files change - such as if you update the version in package.json and run npm-install again
grunt build

# This processes only styles and scripts of the theme only - no Bootstrap files.
grunt theme

# The default task is configured the same as the 'grunt theme' task
grunt

# This deals with building a WPORG distributable version of the theme - it copies only the files required and no dev files or directories. The built theme will be found inside the 'dist/' folder.
grunt dist
```

## Working Directories

The styles and scripts to be modified can be found inside the `assets/src` dirctory. This is the working dirctory and are used as the sourcefiles to build the final versions used by the theme. Final versions are output to `assets/css` and `assets/js` and should not be directly modified in development.

## Build Process Overview

Grunt is used to perform many actions that automate several useful tasks for theme development. The base systems use configurations that emulate closely what BootStrap 4 uses in their build systems. That way I can ensure that the processed Bootstrap files give the same results as the ones available through their pre-built package. It's also a well tested and though through build process.

### Styles:

Both the theme styles and Bootstrap styles use SASS as the pre-processessor. They are processed into 2 seporate files - the Bootstrap stylesheet and the theme stylesheet.

Once the files are turned into CSS postcss is used to apply flex fixes for IE. Finally it uses autoprefixer to make the styles cross-browser compatable.

The autoprefixer config is as follows:

```
browsers: [
  'Chrome >= 35',
  'Firefox >= 38',
  'Edge >= 12',
  'Explorer >= 10',
  'iOS >= 8',
  'Safari >= 8',
  'Android 2.3',
  'Android >= 4',
  'Opera >= 12'
]
```
### Scripts:

Bootstrap scripts are processed with babel (for es2015-module-strip support) then they are combined and minified.

Theme scripts are combined and minified from assets/src/js directory.

## Theme Versioning:

The Version number only needs set in `package.json`. The `dist` task handles updating the version number in style.css and assets/css/style.css during it's copy process.

In development the version is set to `{{ VERSION }}`, the dist package gets updated with correct version numbers.


changelog
2.3.0	- New option to override theme slider and use a shortcode in it's place instead.
2.2.0	- Updated to Bootstrap 4.1.1
		- Added cookies-consent field and styles
2.1.0 	- New option to select number of posts appearing in the loop on frontpage.
		- New option to select category for the frontpage posts loop.
		- New option to put a height cap on the slider to prevent content from being displaced when various height images are used.
		- New options to turn on/off all the rows on the frontpage.
		- Update: Updated wp-bootstrap-navwalker to v4.0.2.
2.0.0	- Moved to Bootstrap 4.0.0 final.
1.7.0	- A lot of developer and child theme friendly focused changes. All relevant functions are now overridable or filterable.
			- Moved default settings to a filterable array.
			- Made all options helper functions filterable.
		- Made the post end widget area output like expected.
		- Added a few more options to the search button color choices.
1.6.0	- Added a small upsell block in customizer.
		- Fix: correctly copy font-awesome styles during `grunt dist`.
		- Enhancement: Added a link on the slide title.
		- Fix: Corrected post-opening-widget section styles in medium sized screens and when in reverse col mode.
		- Fix: Make small title toggle work correctly.
		- Fix: Translate some more strings.
1.5.0	- Updated to Bootstrap@4.0.0.beta.2
		- Updated Popper to 1.12.5
		- Swapped from grunt-contrib-sass to grunt-sass
			- contrib was throwing invalid css warnings at bootstraps new method of adding color variables in root for post-processing.
1.4.4	- Fix: Breadcrumbs wrapped in a proper container.
1.4.3	- Enhancement: no longer a capped amount of widgets in various rows.
		- Accessibility: Added skip link, some aria values, roles.
		- Accessibility: Added text describers for header and footer navs.
		- Dev: WP-Bootstrap-Navwalker has early header/diver support added.
1.4.0	- Added a 'slim mode' for the Bootstrap Styles and Scripts. Includes only what is needed by the theme. Big savings on JS.
			- bootstrap.min.js  =  51.4kb  :  bootstrap-slim.min.js  =  16.2kb
			 	- ~35kb saving or a 68% saving.
			- bootstrap.min.css = 141.8kb  :  bootstrap-slim.min.css = 130.4kb
				- ~10kb saving or an 8% saving.
1.3.0	- Fix: Swapped to front-page.php for front template, instead of a page template.
		- Enhancement: Added Gutenberg wide image support and a color swatch.
		- Enhancement: Added a hook and action for outputting post meta.
		- Fix: Set default value so that intro_text would output.
		- Enhancement: Moved all theme options into own panel and vastly improved explainer text.
		- Tweak: Exclude sticky posts from main home loop to prevent throwing off counts.
		- Enhancement: Add layout select, left/right sidebars.
		- Dev: Improved naming of sidebars.
		- Fix: Swap from `label` to `badge` for post tags class outputs.
		- Enhancement: Add font-awesome.
		- Enhancement: Many new style options for navbar.
			- position: top, bottom, sticky
			- bg colors and text color
		- Enhancement: New options for slider.
			- option to set category used, number of slides.
		- Tweak: Many smaller tweaks, changes and reworks.
1.2.1	- Fix: No background color for body.
		- Fix: Navbar function instead of variable bug.
1.2.0	- Add options for outputting image and site-title as .navbar-brand.
		- Add ability to toggle navbar search form on and off.

1.1.0	- Swap to Bootstrap@4.0.0-beta.
		- Swap out Tether for Popper as new dependency.
		- Accessibility tweak for navwalker.
		- New `.btn-theme` and `.btn-outline-theme`.
		- Fix missing comments ability for pages.

1.0.1	- Fix topbar nav to depth arg of 2 instead of 0 to fix dropdowns.
		- Remove right alignment on topbar toggler button.

1.0.0 	- Fix bootstrap.min.js enqueue
		- Fix image reference in styles for thumb
		- Fix all main theme php files to fit WordPress PHPCS standards.
		- Fix theme mods don't need prefixes - stored in prefixed array
		- Enhancement - swap to using filters and actions to build featured bar
		- Enhancement - Added travis.ci testing for php and a ruleset for phpcs

0.14.0 = WordPress.org Debut

0.14.0	- Fixed styles for gravity forms
		- Fixed translation bug with strings containing sadface - ':('
		- Set content_width as global
		- Translated aria-label
		- Provide dynamic footer default taxt
		- Rework enqueue scripts/styles
		- Remove some unneeded escapes
		- Some pagination changes and styles
		- Remove an unnecessary filters from sanitization functions

0.12.0	- Translation fixes
		- Prefix all functions
		- Correctly reset queries
		- Rework grunt dist task
			- Fix missing script in dist - tether.io
			- Add images to dist
		- Use main site stated time format whenever times are output

0.10.0	- Revert screenshot to a correct sized one

0.10.0	- Translation fixes
		- Remove IE7 or earlier message
		- Add grunt task to generate a clean WPORG version - no dev files

0.9.2	- Translation and text domain fixes

0.8.2 	- Native custom background support added with add_theme_support()
		- Fix markup for slider when no slides are set
		- Moved a closing </div> tag into same file as it was opened
		- Bring forward a change to wp_bootstrap_navwalker
			- Translated string in fallback function
		- Wrap navbar contents inside .container
		- Some default styles and functions added for GF forms
		- Dropped back compatability code for wp_title_tag()
			- Not allwed in theme directory
		- Stop overriding global['comment'] unneccesarily

0.8.1	- Updated theme thumbnail with max-sized screenshot of theme at activation time

0.8 	- Moved to using customizer instead of theme options framework
		- Removed old Twitter code from footer
		- Removed several old files and directories no longer in use
		- Fixed text domain throughout theme to 'best-reloaded'
		- Removed depretiated tags from stylesheets and readme
		- Updated readme with content in-keeping with WP Theme Directory styling
		- Tweaks to custom navwalker to work better with Bootstrap 4.0.0-alpha.6


Gruntfile.js
module.exports = function(grunt) {

	require( 'jit-grunt' )( grunt );

	grunt.initConfig({

		pkg: grunt.file.readJSON( 'package.json' ),

		sass: {

			theme: {
				options: {
					style: 'expanded',
				},
				files: {
					// the first path is the output and the second is the input.
					'assets/src/css/style.css': 'assets/src/scss/style.scss'
				}
			},

			build: {
				options: {
					style: 'expanded',
				},
				files: {
					// the first path is the output and the second is the input.
					'assets/src/css/bootstrap.css'		: 'assets/src/scss/bootstrap/bootstrap.scss',
					'assets/src/css/bootstrap-slim.css'	: 'assets/src/scss/bootstrap/bootstrap-slim.scss',
					'assets/src/css/font-awesome.css' 	: 'assets/src/scss/font-awesome/font-awesome.scss',
					'assets/src/css/style.css' 			: 'assets/src/scss/style.scss'
				}
			}

		},

		postcss: {
			options: {
				map: false,
				processors: [
					require( 'postcss-flexbugs-fixes' ),
					require( 'autoprefixer' )( {
						browsers: [
						  //
						  // Official browser support policy:
						  // https://v4-alpha.getbootstrap.com/getting-started/browsers-devices/#supported-browsers
						  //
						  'Chrome >= 35', // Exact version number here is kinda arbitrary
						  // Rather than using Autoprefixer's native "Firefox ESR" version specifier string,
						  // we deliberately hardcode the number. This is to avoid unwittingly severely breaking the previous ESR in the event that:
						  // (a) we happen to ship a new Bootstrap release soon after the release of a new ESR,
						  //     such that folks haven't yet had a reasonable amount of time to upgrade; and
						  // (b) the new ESR has unprefixed CSS properties/values whose absence would severely break webpages
						  //     (e.g. `box-sizing`, as opposed to `background: linear-gradient(...)`).
						  //     Since they've been unprefixed, Autoprefixer will stop prefixing them,
						  //     thus causing them to not work in the previous ESR (where the prefixes were required).
						  'Firefox >= 38', // Current Firefox Extended Support Release (ESR); https://www.mozilla.org/en-US/firefox/organizations/faq/
						  // Note: Edge versions in Autoprefixer & Can I Use refer to the EdgeHTML rendering engine version,
						  // NOT the Edge app version shown in Edge's "About" screen.
						  // For example, at the time of writing, Edge 20 on an up-to-date system uses EdgeHTML 12.
						  // See also https://github.com/Fyrd/caniuse/issues/1928
						  'Edge >= 12',
						  'Explorer >= 10',
						  // Out of leniency, we prefix these 1 version further back than the official policy.
						  'iOS >= 8',
						  'Safari >= 8',
						  // The following remain NOT officially supported, but we're lenient and include their prefixes to avoid severely breaking in them.
						  'Android 2.3',
						  'Android >= 4',
						  'Opera >= 12'
						]
					} )
				]
			},
			build: {
				files: {
					'assets/css/bootstrap.css'		: 'assets/src/css/bootstrap.css',
					'assets/css/bootstrap-slim.css'	: 'assets/src/css/bootstrap-slim.css',
					'assets/css/font-awesome.css'	: 'assets/src/css/font-awesome.css',
					'assets/css/style.css'			: 'assets/src/css/style.css'
				}
			},
			buildminify: {
				options:{
					sourceMap:false,
					processors: [
						require( 'postcss-clean' )
					]
				},
				files: {
					'assets/css/bootstrap.min.css' 		: 'assets/css/bootstrap.css',
					'assets/css/bootstrap-slim.min.css'	: 'assets/css/bootstrap-slim.css',
					'assets/css/font-awesome.min.css'	: 'assets/css/font-awesome.css',
					'assets/css/style.min.css' 			: 'assets/src/css/style.css'
				}
			},
			theme: {
				files: {
					'assets/css/style.css': 'assets/src/css/style.css'
				}
			},
			thememinify: {
				options:{
					sourceMap:false,
					processors: [
						require( 'postcss-clean' )
					]
				},
				files: {
					'assets/css/style.min.css': 'assets/css/style.css'
				}
			},
		},

		// configure a babel.
		babel: {
			build: {
				options: {
					sourceMap: true,
					presets: [
						[
							"es2015",
							{
								"modules": false,
								"loose": true
							}
						]
					],
					"plugins": [
						"transform-es2015-modules-strip",
						"transform-es2015-destructuring",
					    "transform-es2015-parameters",
					    "transform-object-rest-spread"
					]
				},
				files: {
					'assets/js/util.js'      : 'assets/src/js/bootstrap/util.js',
					'assets/js/alert.js'     : 'assets/src/js/bootstrap/alert.js',
					'assets/js/button.js'    : 'assets/src/js/bootstrap/button.js',
					'assets/js/carousel.js'  : 'assets/src/js/bootstrap/carousel.js',
					'assets/js/collapse.js'  : 'assets/src/js/bootstrap/collapse.js',
					'assets/js/dropdown.js'  : 'assets/src/js/bootstrap/dropdown.js',
					'assets/js/modal.js'     : 'assets/src/js/bootstrap/modal.js',
					'assets/js/scrollspy.js' : 'assets/src/js/bootstrap/scrollspy.js',
					'assets/js/tab.js'       : 'assets/src/js/bootstrap/tab.js',
					'assets/js/tooltip.js'   : 'assets/src/js/bootstrap/tooltip.js',
					'assets/js/popover.js'   : 'assets/src/js/bootstrap/popover.js',
					'assets/js/popper.js'    : 'assets/src/js/popper/popper.js'
				}
			},
			dist: {
				options: {
					presets: [
						[
							"es2015",
							{
								"modules": false,
								"loose": true,
							}
						]
					],
					"plugins": [
						"transform-es2015-modules-strip"
					]
				},
				files: {
					'<%= concat.bootstrap.dest %>' 		: '<%= concat.bootstrap.dest %>',
					'<%= concat.bootstrap_slim.dest %>' : '<%= concat.bootstrap_slim.dest %>',
					'assets/js/popper.js'				: 'assets/js/popper.js'
				}
			}

		},

		concat: {
			options: {
				// Custom function to remove all export and import statements.
				process: function (src) {
					return src.replace( /^(export|import).*/gm, '' )
				}
			},
			bootstrap: {
				src: [
				  'assets/js/util.js',
				  'assets/js/alert.js',
				  'assets/js/button.js',
				  'assets/js/carousel.js',
				  'assets/js/collapse.js',
				  'assets/js/dropdown.js',
				  'assets/js/modal.js',
				  'assets/js/scrollspy.js',
				  'assets/js/tab.js',
				  'assets/js/tooltip.js',
				  'assets/js/popover.js'
				],
				//dest: 'dist/js/<%= pkg.name %>.js'
				dest: 'assets/js/bootstrap.js'
			},
			bootstrap_slim: {
				src: [
				  'assets/js/util.js',
				  'assets/js/button.js',
				  'assets/js/carousel.js',
				  'assets/js/dropdown.js'
				],
				//dest: 'dist/js/<%= pkg.name %>.js'
				dest: 'assets/js/bootstrap-slim.js'
			},
			popper: {
				src: [
					'assets/js/popper.js',
				],
				dest: 'assets/js/popper.js'
			}
		},

		// configure an uglify task.
		uglify: {
			dev: {
				files: [
					{ src: 'assets/js/bootstrap.js', dest: 'assets/js/bootstrap.min.js' }, // All the Bootstrap JS
					{ src: 'assets/js/bootstrap-slim.js', dest: 'assets/js/bootstrap-slim.min.js' }, // Slim Bootstrap JS
					{ src: 'assets/js/popper.js', dest: 'assets/js/popper.min.js' }, // Popper.
				]
			},
			theme: {
				files: [
					{ src: 'assets/js/scripts.js', dest: 'assets/js/scripts.min.js'}
				]
			}

		},

		// Bootstrap, popper and font-awesome are added to the devDependencies
		//  and installed with npm, copy the appropriate files to the /src/ folder.
		copy: {
			build: {
				files: [
					{expand: true, cwd: 'node_modules/bootstrap/scss/', src: ['**'], dest: 'assets/src/scss/bootstrap/'},
					{expand: true, cwd: 'node_modules/bootstrap/js/src/', src: ['**'], dest: 'assets/src/js/bootstrap/', filter: 'isFile'},
					{src: ['node_modules/popper.js/dist/popper.js'], dest: 'assets/src/js/popper/popper.js'},
					{expand: true, cwd: 'node_modules/font-awesome/scss/', src: ['**'], dest: 'assets/src/scss/font-awesome/'},
					{expand: true, cwd: 'node_modules/font-awesome/fonts/', src: ['**'], dest: 'assets/fonts/'},
					{src: ['assets/src/js/scripts.js'], dest: 'assets/js/scripts.js', filter: 'isFile'},
				],
			},
			theme: {
				files: [
					{src: ['assets/src/js/scripts.js'], dest: 'assets/js/scripts.js', filter: 'isFile'},
				],
			},
			// configuration to get the WPORG friendly verion - IE no unneeded files.
			dist: {
				files: [
					// only the needed files from root
					// style.css is copied seporately.
				{
					expand: true,
					src: ['*.php', 'changelog', 'readme.txt', 'screenshot.png'],
					dest: 'dist/best-reloaded/'
				},
				// all files inside /inc/.
				{
					expand: true,
					src: ['inc/**'],
					dest: 'dist/best-reloaded/'
				},
				// only css files - no maps
				// note the unminifide style.css file is copied seporately.
				{
					expand: true,
					src: ['assets/css/style.min.css', 'assets/css/bootstrap.css', 'assets/css/bootstrap.min.css', 'assets/css/bootstrap-slim.css', 'assets/css/bootstrap-slim.min.css', 'assets/css/font-awesome.css', 'assets/css/font-awesome.min.css'],
					dest: 'dist/best-reloaded/'
				},
				// only combined scripts, no individuals.
				{
					expand: true,
					src: ['assets/js/scripts.js', 'assets/js/scripts.min.js', 'assets/js/bootstrap.js', 'assets/js/bootstrap.min.js', 'assets/js/bootstrap-slim.js', 'assets/js/bootstrap-slim.min.js', 'assets/js/popper.js', 'assets/js/popper.min.js'],
					dest: 'dist/best-reloaded/'
				},
				// copy img directory.
				{
					expand: true,
					src: ['assets/img/**'],
					dest: 'dist/best-reloaded/'
				},
				// copy fonts directory.
				{
					expand: true,
					src: ['assets/fonts/**'],
					dest: 'dist/best-reloaded/'
				}],
			},
			addJQueryRemapAtTop: {
				files:[
				// need to remap $ to jQuery in these files.
				{
					expand: true,
					src: ['assets/js/bootstrap.js', 'assets/js/bootstrap-slim.js'],
				}],
				options: {
					process: function (content, srcpath) {
						console.log( 'processing' );
						return 'var $ = jQuery;\n' + content;
					},
				},
			},
			versionReplace: {
				files:[
				// copy stylesheets and insert version number from package.json.
				{
					expand: true,
					src: ['style.css', 'assets/css/style.css'],
					dest: 'dist/best-reloaded/'
				}],
				options: {
					process: function (content, srcpath) {
						console.log( 'processing' );
						var pkgVersion = grunt.file.readJSON( 'package.json' ).version;
						return content.replace( '{{ VERSION }}', pkgVersion );
					},
				},
			}

		},
		watch: {
			styles: {
				files: ['assets/scss/**/*.scss'], // which files to watch.
				tasks: ['default'],
				options: {
					nospawn: true
				}
			},
		}

	});

	grunt.registerTask( 'default', ['sass:theme', 'postcss:theme', 'postcss:thememinify', 'copy:theme', 'uglify:theme'] );
	grunt.registerTask( 'build', ['copy:build', 'sass:build', 'postcss:build', 'postcss:buildminify', 'babel:build', 'concat', 'babel:dist', 'copy:addJQueryRemapAtTop', 'uglify:dev'] );
	grunt.registerTask( 'theme', ['sass:theme', 'postcss:theme', 'postcss:thememinify', 'copy:theme', 'uglify:theme'] );
	grunt.registerTask( 'dist', ['copy:dist', 'copy:versionReplace'] );
};
